cognite_toolkit/_cdf_tk/client/api/extended_data_modeling.py

```diff
@@ -8,3 +27,170 @@ class ExtendedDataModelingAPI(DataModelingAPI):
+            skip_on_version_conflict (bool): If existingVersion is specified on any of the nodes/edges in the input, the default behaviour is that the entire ingestion will fail when version conflicts occur. If skipOnVersionConflict is set to true, items with version conflicts will be skipped instead. If no version is specified for nodes/edges, it will do the writing directly.
+            replace (bool): How do we behave when a property value exists? Do we replace all matching and existing values with the supplied values (true)? Or should we merge in new values for properties together with the existing values (false)? Note: This setting applies for all nodes or edges specified in the ingestion call.
+
+        Returns:
+            InstancesApplyResult: Created instance(s)
+
+        """
+        parameters = {
+            "autoCreateStartNodes": auto_create_start_nodes,
+            "autoCreateEndNodes": auto_create_end_nodes,
+            "autoCreateDirectRelations": auto_create_direct_relations,
+            "skipOnVersionConflict": skip_on_version_conflict,
+            "replace": replace,
+        }
+        tasks = [
+            (self._RESOURCE_PATH, task_items, ToolkitRetryTracker(self._http_client_with_retry.config))
+            for task_items in self._prepare_item_chunks(items, self._CREATE_LIMIT, parameters)
+        ]
+        summary = execute_tasks(
+            self._post_with_item_reduction_retry,
+            tasks,
+            max_workers=min(self._config.max_workers, DATA_MODELING_MAX_WRITE_WORKERS),
+            executor=ToolkitConcurrencySettings.get_data_modeling_write_executor(),
+        )
+
+        def unwrap_element(el: T) -> InstanceApply | T:
+            if isinstance(el, dict):
+                return type(items[0])._load(el, cognite_client=self._cognite_client)
+            else:
+                return el
+
+        def str_format_element(el: T) -> str | T:
+            if isinstance(el, InstanceApply):
+                return f"{el.space}:{el.external_id}"
+            return el
+
+        summary.raise_compound_exception_if_failed_tasks(
+            task_unwrap_fn=lambda task: task[1]["items"],
+            task_list_element_unwrap_fn=unwrap_element,
+            str_format_element_fn=str_format_element,
+        )
+        created_resources = summary.joined_results(lambda res: res.json()["items"])
+
+        return InstancesApplyResultList._load(created_resources)
+
+    def _post_with_item_reduction_retry(
+        self,
+        url_path: str,
+        json_payload: dict[str, Any],
+        retry_tracker: ToolkitRetryTracker,
+    ) -> Response:
+        _, full_url = self._resolve_url("POST", url_path)
+        headers = self._configure_headers(
+            "application/json",
+            additional_headers=self._config.headers.copy(),
+        )
+        data: str | bytes
+        try:
+            data = _json.dumps(json_payload, allow_nan=False)
+        except ValueError as e:
```

Comment:
Are we sure this is the only thing that can throw `ValueError` here?

Reply:
Yes, `JSONDecodeError` is a subclass of ValueError.

Reply:
I guess one risk here is the error message changing, and users installing this lib with a newer version of the dependency from what is tested here.

Reply:
That seems like a very small risk that the builtin `json` module stops making `JSONDecodeError` a subclass of `ValueError`.

Reply:
but if that does happen, it only means we return a more generic exception